### PR TITLE
Json format application logs for kibana

### PR DESCRIPTION
**Pull request by mburke5678**
https://bugzilla.redhat.com/show_bug.cgi?id=1710864

Steps taken from:
https://richm.github.io/add-fields-to-index-template
https://richm.github.io/how-to-add-fields-to-kibana

### CODE REVIEW

File: install_config/aggregate_logging.adoc [hunk 1]
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+					"type": "date"
+				}
+			}
+		}
+	},
+	"template": "project.<project-name>.*" <1>
+}
+----
+<1> Specify the `.operations.*` or a `project.` index pattern.
+
```
Comment (mburke5678) [inline install_config/aggregate_logging.adoc:2105]: @QiaolingTang 
What would the user put as `template:` for the OPS cluster index pattern?

File: install_config/aggregate_logging.adoc [hunk 2]
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+. Get the name of one of the Elasticsearch pods:
++
+----
+$ oc get -n logging pods -l component=es
+
+NAME                                       READY     STATUS    RESTARTS   AGE       IP             NODE                         NOMINATED NODE
+logging-es-data-master-5av030lk-1-2x494    2/2       Running   0          38m       154.128.0.80   ip-153-12-8-6.wef.internal   <none>
+----
+
```
Comment (mburke5678) [inline install_config/aggregate_logging.adoc:2114]: @QiaolingTang @richm 
For a project index pattern, which pod to use?

File: install_config/aggregate_logging.adoc [hunk 3]
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+  for ff in $( oc exec -c elasticsearch <espod-name> -- ls /usr/share/elasticsearch/index_patterns ) ; do
+    oc exec -c elasticsearch <es-pod-name> -- cat /usr/share/elasticsearch/index_patterns/$ff > $ff
+  done
+  break
+done
+----
++
+The index patterns download to the *_/usr/share/elasticsearch/index_patterns_*
+directory.
+
```
Comment (mburke5678) [inline install_config/aggregate_logging.adoc:2142]: @QiaolingTang @richm 
Does anyone have a  list of index pattern file names?

File: install_config/aggregate_logging.adoc [hunk 4]
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+. Load the JSON file into the Elasticsearch pod:
++
+----
+$ cat <json-file-name> | \
+oc exec -n logging -i -c elasticsearch <es-pod-name> -- \
+    curl -s -k --cert /etc/elasticsearch/secret/admin-cert \
+    --key /etc/elasticsearch/secret/admin-key \
+    https://localhost:9200/_template/<json-file-name> -XPUT -d@- | \
+python -mjson.tool
+----
```
Comment (mburke5678) [inline install_config/aggregate_logging.adoc:2124]: @QiaolingTang @richm 
What would this command look like to add a JSON to a pod if using project index

File: install_config/aggregate_logging.adoc [hunk 5]
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+$ for espod in $( oc get pods -l component=es -o jsonpath='{.items[*].metadata.name}' ) ; do
+  for ff in $( oc exec -c elasticsearch <esops-pod-name> -- ls /usr/share/elasticsearch/index_patterns ) ; do
+    oc exec -c elasticsearch <esops-pod-name> -- cat /usr/share/elasticsearch/index_patterns/$ff > $ff
+  done
+  break
+done
+----
++
+The index patterns are download to the *_/usr/share/elasticsearch/index_patterns_* directory.
+
```
Comment (mburke5678) [inline install_config/aggregate_logging.adoc:2185]: Would the user need to do this for the OPS cluster or would the index patterns be the same?

File: install_config/aggregate_logging.adoc [hunk 6]
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+The data type must correspond to the Elasticsearch field definition you added above. For example, if you added a field `myfield` in Elasticsearch that is a `number` or `integer` type, you cannot add the `myfield` to Kibana as a `string` type.
+
+. Add the name of the Kibana index pattern to the index pattern files:
++
+For example, to use the *project.MYNAMESPACE.* index pattern:
++
+----
+"title": "project.MYNAMESPACE.*"
+----
+
```
Comment (mburke5678) [inline install_config/aggregate_logging.adoc:2207]: @QiaolingTang @richm 
For operations, the user would put just `"title": "operations.*"`?

**Comment by mburke5678** [comment]
@QiaolingTang @richm I added some questions in-line.  PTAL.

@QiaolingTang If you have a 3.11 cluster with logging installed that I could look at, I might be able to answer some of my questions.

**Comment by mburke5678** [comment]
@richm Can you help me with this? I am trying to add your blogs on adding custom fields to Kibana to the user docs, at customer request. I don't have access to a 3.11 cluster with logging. I tried the steps in 4.4, but it doesn't work (Different Elastic version?)

https://richm.github.io/add-fields-to-index-template
https://richm.github.io/how-to-add-fields-to-kibana

**Comment by richm** [comment]
> @richm Can you help me with this? I am trying to add your blogs on adding custom fields to Kibana to the user docs, at customer request. I don't have access to a 3.11 cluster with logging.

I have a 3.11 system you can use.

> I tried the steps in 4.4, but it doesn't work (Different Elastic version?)

What exactly is not working?

> 
> https://richm.github.io/add-fields-to-index-template
> https://richm.github.io/how-to-add-fields-to-kibana

**Comment by mburke5678** [comment]
> What exactly is not working?
I created the JSON file. 
Added to ES:
```
cat $file | oc exec -n openshift-logging -i -c elasticsearch $espod -- curl -s -k --cert /etc/elasticsearch/secret/admin-cert --key /etc/elasticsearch/secret/admin-key https://localhost:9200/_template/$file -XPUT -d@- | python -mjson.tool
{
    "acknowledged": true
}
```
Ran the command to verify:

```
oc exec -n openshift-logging -i -c elasticsearch $espod -- curl -s -k --cert /etc/elasticsearch/secret/admin-cert --key /etc/elasticsearch/secret/admin-key https://localhost:9200/testy2.*/_search?sort=mytimefield2:desc | python -mjson.tool
{
    "_shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 0,
        "total": 0
    },
    "hits": {
        "hits": [],
        "max_score": 0.0,
        "total": 0
    },
    "timed_out": false,
    "took": 0
} 
```
(Don't like the looks of all those zeros)
Added to Kibana, but am not seeing the custom field mytimefield2

**Comment by richm** [comment]
couple of points:

> NOTE The settings will not apply to existing indices. You will need to perform a reindexing for that to work. However, it is usually not a problem, as the settings will apply to new indices

New indices are created at midnight or so, by default.  I suppose one way you could do it would be to create a new namespace, and create a new pod in that namespace, and have something running in that pod that produces JSON output with your new field in the JSON, then configure `MERGE_JSON_LOG=true` so that fluentd will output the field.

> To see if this is working, wait until new indices are created,

See above.

File: install_config/aggregate_logging.adoc [hunk 7]
```diff
@@ -2054,6 +2054,244 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
 
 Where `$espod` is the name of any one of your Elasticsearch pods.
 
+[[kibana-visualizations-custom]]
+*Adding Custom Fields to Kibana Visualize*
+
+If your {product-title} cluster is configured with custom logging fields
```
Comment (QiaolingTang) [inline install_config/aggregate_logging.adoc:2060]: What about `If your {product-title} cluster generates logs in json format, which consist of custom fields` ?

File: install_config/aggregate_logging.adoc [hunk 8]
```diff
@@ -2054,6 +2054,244 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+that are not are not defined in the `.operations.\*` or the `project.*` 
+index, you cannot create visualizations with these fields. 
+
+However, if you if you use the `MERGE_JSON_LOG` feature (`MERGE_JSON_LOG=true`) 
+to generate JSON format logs, you can use these custom fields in
```
Comment (QiaolingTang) [inline install_config/aggregate_logging.adoc:2065]: I think we don't need to mention `MERGE_JSON_LOG` here as this value is set to `true` by default.

File: install_config/aggregate_logging.adoc [hunk 9]
```diff
@@ -2054,6 +2054,244 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
...
+that are not are not defined in the `.operations.\*` or the `project.*` 
+index, you cannot create visualizations with these fields. 
+
+However, if you if you use the `MERGE_JSON_LOG` feature (`MERGE_JSON_LOG=true`) 
+to generate JSON format logs, you can use these custom fields in 
+Kibana Visualize by adding the fields to the Elasticsearch index patterns and to Kibana. 
+
+To add custom fields to Kibana Visualize:
+
+. Add custom fields to an Elasticsearch index template:
```
Comment (QiaolingTang) [inline install_config/aggregate_logging.adoc:2070]: We may need to add a note here: these fields will only be applied to the indices created after the template is updated.

**Comment by mburke5678** [comment]
@QiaolingTang I made the changes as you suggested. PTAL

**Comment by QiaolingTang** [comment]
lgtm

**Comment by mburke5678** [comment]
@QiaolingTang Thank you for the quick review!

File: install_config/aggregate_logging.adoc [hunk 10]
```diff
@@ -2054,6 +2054,252 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
 
 Where `$espod` is the name of any one of your Elasticsearch pods.
 
+[[kibana-visualizations-custom]]
+*Adding Custom Fields to Kibana Visualize*
+
+If your {product-title} cluster generates logs in JSON format that contain custom fields
+that are not are not defined in the Elasticsearch `.operations.\*` or the `project.*`
```
Comment (lamek) [inline install_config/aggregate_logging.adoc:2061]: Extra words here, "are not **are not**"

**Pull request merged**